### ISSUE TITLE
modules used on a worker thread are not cached in indexedDB

### ISSUE THREAD

**Issue by benjie-git** [brython-dev/brython#2135]
As far as I can tell, the modules used on a worker thread are not cached into the indexedDB.  Is this a known limitation?  In my case, most of my python code is running in a worker, so I expect my load times would improve with indexedDB caching of worker thread modules.

**Comment by PierreQuentel** [comment]
Good suggestion Ben ! Currently this is not implemented because indexedDB is asynchronous, I'm not sure if it's possible to use it in a web worker without changing the whole worker API. I will work on this issue in the next days.

**Comment by PierreQuentel** [comment]
In the commit referenced above I have implemented a new API to create a web worker that can take advantage of the IndexedDB cache. It is described in the new version of the [documentation](https://github.com/brython-dev/brython/blob/master/www/doc/en/browser.worker.md).

Could you test it, tell me if you find bugs and what you think of the API design ?

**Comment by benjie-git** [comment]
Thank you!  I tested the
   <script type="text/python" class="webworker" id="myworker" src="myworker.py"></script>
method of creating a worker, and it looks like it's working great, and all modules are being added to the indexedDB.  I haven't yet tested that these are actually being used on subsequent loads, but I'd expect that they are!  I'll try to test this further in a few days when I'm back home.

Is there any benefit to calling create_worker() from the main brython thread instead?  I imagine that in that case the worker couldn't begin loading until after the main brython code starts running, whereas the <script> tag method could potentially start loading the main thread and the worker in parallel.  Do you know if that actually happens?

**Comment by PierreQuentel** [comment]
By design, workers are started by a "main program". I suppose that it's because they are liable to send data to their "caller" when they are loaded, and if the main program was not active at that time this data would be lost.

In all the examples I have seen, written in Javascript, the worker is started by `new Worker(worker_program.js)`. The ability to have the worker code in the same page as the main program is specific to Brython.

**Comment by benjie-git** [comment]
I'm seeing my pages load quite a bit faster now.  Thank you again!

**Comment by benjie-git** [comment]
I forgot to comment on the API.  After sitting down with it again, I realized that my previous reply was a bit nonsensical.  Sorry about that!  I fully transitioned to using the new API, and I like that it is asynchronous, and that the create_worker() call includes separate callbacks for onready and onerror.  (Do you want the python arg names to be on_ready, on_error, and on_message?)

At first I wasn't sure about including onmessage in the create_worker() call, as it seems like something that could make more sense to only set up in onready, only if/when the worker is set up without error, but I realized that setting it up in the create_worker() call ensures that we don't miss any messages sent by the worker immediately upon creation, so I think the way you have it designed currently is excellent.

Thanks again!

**Comment by PierreQuentel** [comment]
Thanks for the feedback. I hesitated between `on_ready` and `onready`, but in **`browser.ajax`** the callbacks to `ajax.get()` and the like don't have the underscore, so I kept the underscore-less version for consistency.

I will publish a message on the Discussions section to see if there are other opinions.
